internal/util/util.go

```diff
@@ -530,3 +536,100 @@ func HttpRequest(method string, url string, headers []string) (resp *http.Respon
 	}
 	return client.Do(req)
 }
+
+func resolvePath(filename string) (string, error) {
+	for {
+		// We use Stat() as trick to let the OS do symlink loop detection for us,
+		// since we are lazy to do it ourselves
+		_, err := os.Stat(filename)
+		if errors.Is(err, fs.ErrNotExist) {
+			return filename, nil
+		}
+		if err != nil {
+			return filename, err
+		}
+		fileInfo, err := os.Lstat(filename)
+		if err != nil {
+			return filename, err
+		}
+		if fileInfo.Mode()&os.ModeSymlink != 0 {
+			dstFilename, err := os.Readlink(filename)
+			if err != nil {
+				return filename, err
+			}
+			if !filepath.IsAbs(dstFilename) {
+				filename = filepath.Join(filepath.Dir(filename), dstFilename)
+				continue
+			}
+		}
+		return filename, nil
+	}
+}
+
+// OverwriteFile opens the given file for writing, truncating if one exists, and then calls
+// the supplied function with the file as io.Writer object, also making sure the file is
+// closed afterwards.
+func OverwriteFile(name string, enc encoding.Encoding, fn func(io.Writer) error, withSudo bool, sucmd string, lock func(), unlock func()) (err error) {
+	var writeCloser io.WriteCloser
+	var cmd *exec.Cmd
+
+	name, err = resolvePath(name)
+	if err != nil {
+		return
+	}
+
+	if withSudo {
+		cmd = exec.Command(sucmd, "dd", "bs=4k", "of="+name)
+
+		if writeCloser, err = cmd.StdinPipe(); err != nil {
+			return
+		}
+
+		c := make(chan os.Signal, 1)
+		signal.Notify(c, os.Interrupt)
+		go func() {
+			<-c
+			cmd.Process.Kill()
+		}()
+
+		lock()
+		// need to start the process now, otherwise when we flush the file
+		// contents to its stdin it might hang because the kernel's pipe size
+		// is too small to handle the full file contents all at once
+		if e := cmd.Start(); e != nil && err == nil {
+			unlock()
+			return err
+		}
+	} else if writeCloser, err = os.OpenFile(name, os.O_WRONLY|os.O_CREATE|os.O_TRUNC, 0666); err != nil {
```

Comment:
I've tried to add some keybinding via the `bind` command (with your PR applied)... I shouldn't have. It said `Backup file doesn't exist` and wiped my `bindings.json`.

I have no idea why it failed to find the backup (and I can't reproduce it again). As to why it truncated the file while it should have left in untouched, I guess, well, at least this `os.O_TRUNC` is the answer...

So, doing all the backup manipulations *inside* the `fn()` callback seems wrong. It seems we should instead just `OverwriteFile()` the backup file first, then `OverwriteFile()` the target file?

Reply:
> at least this os.O_TRUNC is the answer

Damn, yes...didn't care about these flags so far. Do we really need to open it that way? Because according to the docs it should truncate on write anyway or am I wrong?

Anyway, decoupling seems to be better then and most probably easier to follow.

Reply:
It either truncates on open() or doesn't truncate at all. It doesn't truncate on write(). That's the common Unix semantics. Which docs say otherwise?

Reply:
[os#WriteFile](https://pkg.go.dev/os#WriteFile) explains it the following:
> [...]; otherwise WriteFile truncates it before writing, [...]

But I'm unsure if different `go` interfaces treat it differently and if I misunderstood it.

Reply:
`os.WriteFile()` aka `ioutil.WriteFile()` is a helper which does all that at once: it opens the file (truncating it), writes the entire content to it, and closes it.

Why can't we just use this `os.WriteFile()` then? First, because, you know, we have this sudo stuff and this encoding stuff (though that is relevant to buffer files only), and second, because we want to sync the file to the disk (and that is relevant to both buffer and non-buffer files), which we do via `f.Sync()`. With `os.WriteFile()` we can't sync the file, since we don't have its descriptor (so we could only sync everything, not just this file, which is very inefficient).